[Docs] Update key server setup (#158)
MystenLabs/seal: UsingSeal.md

## For key server operators

The key server can run either in `Open` mode or `Permissioned`.
`Open` mode allows querying keys for any onchain package and uses a single IBE master key for all of them.
`Permissioned` mode restricts queries to manually approved packages, associated with a specific client. It also supports importing and exporting a key for the specific client if needed. More on both options in the following.

> ### Open mode

Reviewer: For this section, until the `sui client call` example, I would suggest to change to the following (sorry providing like this as it's hard to provide line by line feedback below):

> In open mode, the key server allows decryption requests for Seal policies from any package. This mode is ideal for testing or for deployments where the key server is operated as a best-effort service without direct user liability.
> 
> **Generate a Master Key**
> 
> Before starting the key server, you must generate a BLS master key pair. Use the `seal-cli` tool as ``cargo run --bin seal-cli genkey`.
> 
> This command outputs both the master secret key and the public key. These keys are used to derive identity-based encryption (IBE) keys during policy enforcement.
> 
> **Register the Key Server On-Chain**
> 
> To make the key server discoverable by Seal clients, register it on-chain.
> Call the `create_and_transfer_v1` function from the `seal::key_server` module like following:

In this mode Seal policies of any package can be queried.
This is an ideal option for testing Seal, or if the key server does not have any direct liabilty to users (*best-effort* service).

First, a master BLS keypair should be generated.
Use the `seal-cli` tool for that using `cargo run --bin seal-cli genkey`.
It outputs the master secret key and the public key.

Key servers should be registered onchain to enable discoverability.
To register a key server, call the `create_and_transfer_v1` function in the `seal::key_server` module. For example:

```shell
sui client call --function create_and_transfer_v1 --module key_server --package 0x62c79dfeb0a2ca8c308a56bde530ccf3846535e1623949d45c90d23128afff52 --args mysten-dev-1 https://seal-key-server-testnet-1.mystenlabs.com 0 0xa023acbf600401017ee17bf918106ea9911914ca017aa3ab9ab5c64beb9bb5236fd9d4d5b5645dc3bc0d4f732ed04fc60d14b9f37987fe5eeb4db07fc0982904ce1ed0b07607ae2e99086e141f6c6a1df6def5f5d434ca7c09856a3750c92969 --gas-budget 10000000
```

> Run the server using `cargo run --bin key-server` with the environment variable 

Reviewer: For this entire statement before the example shell commands, here's a suggested edit:

> To start the key server in open mode, run the command `cargo run --bin key-server`, but before running the server, set the following environment variables:
> 
> - MASTER_KEY: The master secret key generated using the `seal-cli genkey` command.
> - CONFIG_PATH: The path to a .yaml configuration file that specifies key server settings. For the configuration file format, see the [example config](crates/key-server/key-server-config.yaml).
> 
> In the config file, make sure to:
> - Set the mode to `Open`.
> - Set the `key_server_object_id` field to the ID of the key server object you registered on-chain.

`MASTER_KEY` set to the master secret key generated by the `seal-cli` tool,
and the environment variable `CONFIG_PATH` pointing to a .yaml configuration file. See [example](crates/key-server/key-server-config.yaml) for the format of the configuration file and use the `Open` mode. Set `key_server_object_id` to the id of the registered key server object.

> Example:

Reviewer: ```suggestion
For example:
```

```shell
export MASTER_KEY="0x652081a0ad50cd516e8cb82e44653793a27853350d35d724fded523e85ed0382"
export CONFIG_PATH="/Users/alice/seal/key-server-config.yaml"
cargo run --bin key-server
```

> // TODO: update the docker

Reviewer: What does this refer to? Does it need to be updated before being merged?

Author: it's a todo to fix the docker

curl -H 'Content-Type: application/json' -d '{"ptb":"AwAgv7n4Pj/owIGk+nMEXJ4KbXvVytgdurspb4BZJ8RdY6MAISDAbM35ybJGgprxCy0H1MuWKsGvRO5zQfk4jn+oAcufAwEBwGzN+cmyRoKa8QstB9TLlirBr0Tuc0H5OI5/qAHLnwOtCKsRAAAAAAABAG0lMvMLkGCIOUyRinodOfXWx1Bxm0JCFS/6RhBLoEUxCXdoaXRlbGlzdAtrbXNfYXBwcm92ZQADAQAAAQEAAQIA","enc_key":"lR69cNXFuB3zyjDI12syH1XC4sfaTo63Ylms/I8yO39OTvuDmngPln8pMTJEwq9v","request_signature":"Ut6iy+f+l/FeMgcim+s3jkQO86nxgqhWweOfqFWwQGi3gercjBnzhoMmzrb0i0Z79iXUwE56GwM/2mgHDRz+CA==","certificate":{"session_vk":"w+xC1R2fZDJeQW4lERxYsd3XVAYy7wOx5sKsIWdwlWw=","creation_time":1737671115293,"ttl_min":9,"signature":"APDPqXOr+HqHMw48qNwDMPezgvwDzojcIaBztBdOYuTaomY2pWBVyXbaFyNN+huz5f44ZgMG1D6PnF/H/OGgkADgFHRr36qyRWIXLGJ/iArkqpw1lMzlikMeDlBjTOheIw=="}}' -X POST http://0.0.0.0:2024/v1/fetch_key
``` -->

Key servers expose a set of metrics via a Prometheus server running on port 9184. They can be viewed in raw form by calling `curl http://0.0.0.0:9184`
or used in a data visualization and analytics tool like Grafana.

Similarly, an health check endpoint can be used by calling

> `curl http://0.0.0.0:2024/health`.

Reviewer 2: move these two lines to somewhere outside Open mode? should be applied to both modes

### Permissioned mode
In this mode only Seal policies of allowlisted packages can be queried.
This is the recommended option for B2B use cases.

First, a master seed should be generated.
Use the `seal-cli` tool for that using `cargo run --bin seal-cli gen-seed`.
It outputs the secret master seed.

> Create a configuration file with `Permissioned` mode and no clients following the instructions from the [example](crates/key-server/key-server-config.yaml).

Reviewer 2: this file is currently defaults with Open settings, if we use this as example, its not very clear that one should not use this file as is. consider make it clear on what an "example" look like?

Reviewer: Apart from Joy's comment (agree that a additional example config file with Permissioned mode will be useful), would suggest to change the section above to:

> In permissioned mode, the key server only allows decryption requests for Seal policies from explicitly allowlisted packages. This is the recommended mode for B2B deployments where tighter access control and client-specific key separation are required.
> 
> **Generate a Master Seed**
> 
> Start by generating a master seed for the key server.  Use the `seal-cli` tool as `cargo run --bin seal-cli gen-seed`.
> 
> This command outputs the secret master seed, which is used to derive individual IBE keys per allowlisted client. Store the master seed securely.
> 
> **Create the Configuration File**
> 
> Next, create a configuration file in .yaml format with the following properties:
> 
> - Set the mode to `Permissioned`.
> - Initialize with an empty client list (clients can be added dynamically later).
> 
> You can refer to the [example configuration file](crates/key-server/key-server-config.yaml) for the expected structure and field definitions.

Author: >this file is currently defaults with Open settings, if we use this as example, its not very clear that one should not use this file as is. consider make it clear on what an "example" look like?

The file includes example of Permissioned in comments as well, what else do you think we should add/document?

Run the server using `cargo run --bin key-server` with the environment variable 

> `MASTER_SEED` set to the master secret seed generated by the `seal-cli` tool,

Reviewer 2: you mean set MASTER_KEY..? i dont see MASTER_SEED anywhere in code?

Reviewer 2: ```
It should abort after printing a list of unassigned derived public keys (search for logs with the text `Unassigned derived public key`).
```
i didn't understand this part on this first read, i also didnt get to abort - should this be a separate command?

Reviewer: Will wait to suggest once Joy's comment is addressed

Author: >you mean set MASTER_KEY..? i dont see MASTER_SEED anywhere in code?

see the code of server.rs

> i didn't understand this part on this first read, i also didnt get to abort - should this be a separate command?

did you run the code from this branch?

and the environment variable `CONFIG_PATH` pointing to a .yaml configuration file.

> It should abort after printing a list of unassigned derived public keys (search for logs with the text `Unassigned derived public key`).

Reviewer 2: its slightly easier to read if a code block is provided in addition to this description, e.g. 
```
cargo run --bin seal-cli gen-seed
Seed: b2fcdcef1e9f08709b6a1ac1d451f49d00a87ac028dd9fdad924a5c50d62543d # use this value below

export MASTER_SEED=b2fcdcef1e9f08709b6a1ac1d451f49d00a87ac028dd9fdad924a5c50d62543d
CONFIG_PATH=crate/key-server-config.yaml 
```

(its currently failing with LEGACY_OBJECT_ID but can ignore for now)

Reviewer: Will wait to suggest once Joy's comment is addressed

Author: >(its currently failing with LEGACY_OBJECT_ID but can ignore for now)

LEGACY_OBJECT_ID?

> Each supported client requires registering an onchain key server object to enable discoverability.

Reviewer: Suggest the below edit for entire section until before "Infrastructure Requirements":

> Each supported client must have a registered on-chain key server object to enable discovery and policy validation.
> 
> **Register the First Client**
> 
> - Use the first unassigned derived public key (with derivation index `0`).
> - Register a new key server on-chain by calling the `create_and_transfer_v1` function from the `seal::key_server` module (see `Open Mode` above for usage).
> - In your key server configuration file:
>     - Add a new client entry.
>     - Set `client_master_key` to type `Derived` with `derivation_index` as `0`.
>     - Include the list of packages this client will use.
> - Restart the key server to apply the changes, using `cargo run --bin key-server`.
> 
> **Register Additional Clients**
> 
> To add more clients:
> 
> - Check the key server logs for the next unassigned derived public key.
> - Register the new public key on-chain using `create_and_transfer_v1`.
> - Add a new client configuration using the next `derivation_index` value and corresponding packages.
> 
> **Exporting a Client Key**
> 
> In rare cases where you need to export a client key (for example, during key server rotation):
> 
> - Use the seal-cli tool as `cargo run --bin seal-cli derive-key --seed $MASTER_SEED --index X`. Replace `X` with the `derivation_index` of the key you want to export. The tool will output the corresponding BLS master key, which can be imported by another key server if needed.
> - To disable this key on the current server:
>     -  Change the client's `client_master_key` type to `Exported`.
>     - Add the `deprecated_derivation_index` field with the appropriate index.
> 
> **Importing a Client Key**
> 
> To import a BLS master key into a new key server:
> 
> - Set an environment variable with the imported key (e.g., `EXPORTED_CLIENT_KEY`).
> - In the config file, add a new client entry with:
>     - `client_master_key` set to type `Imported`.
>     - The name of the environment variable containing the key.
>     - The packages associated with the client.
>     - The ID of the key server object registered on-chain for this client.
> 
> **Note**: After importing, the registered key server object must be transferred onchain to the new key server’s owner. The new owner should also update the key server’s URL to reflect the new endpoint.

curl -H 'Content-Type: application/json' -d '{"ptb":"AwAgv7n4Pj/owIGk+nMEXJ4KbXvVytgdurspb4BZJ8RdY6MAISDAbM35ybJGgprxCy0H1MuWKsGvRO5zQfk4jn+oAcufAwEBwGzN+cmyRoKa8QstB9TLlirBr0Tuc0H5OI5/qAHLnwOtCKsRAAAAAAABAG0lMvMLkGCIOUyRinodOfXWx1Bxm0JCFS/6RhBLoEUxCXdoaXRlbGlzdAtrbXNfYXBwcm92ZQADAQAAAQEAAQIA","enc_key":"lR69cNXFuB3zyjDI12syH1XC4sfaTo63Ylms/I8yO39OTvuDmngPln8pMTJEwq9v","request_signature":"Ut6iy+f+l/FeMgcim+s3jkQO86nxgqhWweOfqFWwQGi3gercjBnzhoMmzrb0i0Z79iXUwE56GwM/2mgHDRz+CA==","certificate":{"session_vk":"w+xC1R2fZDJeQW4lERxYsd3XVAYy7wOx5sKsIWdwlWw=","creation_time":1737671115293,"ttl_min":9,"signature":"APDPqXOr+HqHMw48qNwDMPezgvwDzojcIaBztBdOYuTaomY2pWBVyXbaFyNN+huz5f44ZgMG1D6PnF/H/OGgkADgFHRr36qyRWIXLGJ/iArkqpw1lMzlikMeDlBjTOheIw=="}}' -X POST http://0.0.0.0:2024/v1/fetch_key
``` -->

### Permissioned mode
In `Permissioned` mode, the key server only allows decryption requests for Seal policies from explicitly allowlisted packages.

> This is the recommended mode for B2B deployments where tighter access control and client-specific key separation are required.

Reviewer: ```suggestion
This is the **recommended mode for B2B deployments** where tighter access control and client-specific key separation are required.
```

Start by generating a master seed for the key server. 
Use the `seal-cli` tool as `cargo run --bin seal-cli gen-seed`.
This command outputs the secret master seed which should be stored securely.

Next, create a configuration file in .yaml format following the instructions in the [example config](crates/key-server/key-server-config.yaml) 
and with the following properties:
- Set the mode to `Permissioned`.

> - Initialize with an empty client configs (clients can be added later).

Reviewer: ```suggestion
- Initialize with an empty client config (see instructions below for adding clients).
```

Set the environment variable `MASTER_SEED` to the master secret seed generated by the `seal-cli` tool,
and the environment variable `CONFIG_PATH` pointing to a .yaml configuration file.
Run the server using `cargo run --bin key-server`.
It should abort after printing a list of unassigned derived public keys (search for logs with the text `Unassigned derived public key`).

> Each supported client must have a registered on-chain key server object to enable discovery and policy validation.

Reviewer: ```suggestion
> [!Note] Each supported client must have a registered on-chain key server object to enable discovery and policy validation.
```

> Register the first client:

Reviewer: ```suggestion
To register the first client:
```

- Use the first unassigned derived public key (with derivation index 0).
- Register a new key server on-chain by calling the `create_and_transfer_v1` function from the `seal::key_server` module (see Open Mode above for usage).
- In your key server configuration file:
  - Add a new client entry.
  - Set `client_master_key` to type `Derived` with `derivation_index` as 0. 
  - Include the list of packages this client will use.
Restart the key server to apply the changes, using `cargo run --bin key-server`.

> To add more clients:

Reviewer: ```suggestion
To add additional clients:
```

- Check the key server logs for the next unassigned derived public key.
- Register the new public key on-chain using `create_and_transfer_v1`.

> - Add a new client configuration using the next `derivation_index` value and corresponding packages.

Reviewer: ```suggestion
- Add a new client configuration using the next `derivation_index` value and corresponding packages.
Restart the key server to apply the changes, using `cargo run --bin key-server`.
```

#### Export / Import Keys

In rare cases where you need to export a client key:
- Use the `seal-cli` tool as `cargo run --bin seal-cli derive-key --seed $MASTER_SEED --index X`. 
Replace `X` with the `derivation_index` of the key you want to export.
The tool will output the corresponding master key, which can be imported by another key server if needed.
- Disable this key on the current server:
  - Change the client's `client_master_key` type to `Exported`.
  - Set the `deprecated_derivation_index` field with the appropriate index.

To import a BLS master key into a new key server:
- Set an environment variable with the imported key (e.g., `EXPORTED_CLIENT_KEY`).
- In the config file, add a new client entry with:
  - `client_master_key` set to type `Imported`.
  - The name of the environment variable containing the key.
  - The packages associated with the client.
  - The ID of the key server object registered on-chain for this client.

> **Note:** After importing, the registered key server object must be transferred onchain to the new key server’s owner.

Reviewer: ```suggestion
> [!Note] After importing, the registered key server object must be transferred onchain to the new key server’s owner.
```
